objects/src/mock/chain.rs

```diff
@@ -55,3 +55,401 @@ pub fn mock_chain_data(consumed_notes: &mut [Note]) -> ChainMmr {
+            self.nonfungible_faucets.extend(pending.nonfungible_faucets.drain(..));
+
+            pending.set_notes_proofs(header, &notes);
+            self.notes.extend(pending.notes.drain(..));
+            pending.nullifiers.clear(); // nullifiers are saved in the nullifier TSTM
+        }
+
+        /// Creates a [SimpleSmt] tree from the `notes`.
+        ///
+        /// The root of this tree is a commitment to [Note]s included in the block. The commitment
+        /// is not for the fields in the [Note] structure, but for the [NoteMetadata] and for a
+        /// subset of the [Note]'s attributes (i.e. recipient + vault).
+        pub fn build_notes_tree(&self) -> Result<SimpleSmt, MerkleError> {
+            let mut entries = Vec::with_capacity(self.notes.len() * 2);
+
+            entries.extend(self.notes.iter().enumerate().map(|(index, note)| {
+                let tree_index = (index * 2) as u64;
+                (tree_index, note.hash().into())
+            }));
+            entries.extend(self.notes.iter().enumerate().map(|(index, note)| {
+                let tree_index = (index * 2 + 1) as u64;
+                (tree_index, note.metadata().into())
+            }));
+
+            SimpleSmt::with_leaves(NOTE_LEAF_DEPTH, entries)
+        }
+
+        /// Given the [BlockHeader] and its notedb's [SimpleSmt], set all the [Note]'s proof.
+        ///
+        /// Update the [Note]'s proof once the [BlockHeader] has been created.
+        fn set_notes_proofs(&mut self, header: BlockHeader, notes: &SimpleSmt) {
+            self.notes.iter_mut().enumerate().for_each(|(index, note)| {
+                let auth_index =
+                    NodeIndex::new(NOTE_TREE_DEPTH, index as u64).expect("index bigger than 2**20");
+                let note_path =
+                    notes.get_path(auth_index).expect("auth_index outside of SimpleSmt range");
+                note.set_proof(
+                    NoteInclusionProof::new(
+                        header.block_num(),
+                        header.sub_hash(),
+                        header.note_root(),
+                        index as u64,
+                        note_path,
+                    )
+                    .expect("Invalid data provided to proof constructor"),
+                );
+            });
+        }
+    }
+
+    /// Structure chain data, used to build necessary openings and to construct [BlockHeader]s.
+    #[derive(Debug, Clone)]
+    #[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+    pub struct MockChain<R: Rng + SeedableRng> {
+        /// An append-only structure used to represent the history of blocks produced for this chain.
+        chain: ChainMmr,
+
+        /// Storage for the historical state of the chain.
+        ///
+        /// This store contains historical data for accounts, nullifiers, notes, and batches.
+        data: DefaultMerkleStore,
```

Comment:
I wonder whether we actually need to store historical data for accounts and nullifiers. I guess it doesn't hurt, but I'm also not sure what we'd use it for.

Regarding notes: since we attach inclusion paths to each note when we seal a block, it seems like this data may be redundant (i.e., if we need to get a note inclusion path, we'd go to the note directly).